### ISSUE TITLE
STATUS_ACCESS_DENIED (0xc0000022): Query directory failed & STATUS_ACCESS_DENIED (0xc0000022): Error closing connection

### ISSUE THREAD

**Issue by himanshuccl** [hierynomus/smbj#737]
Hi Team,

We were upgrading to smbj 0.11.1 (or 0.11.5) - inorder to provision support of SMBv3 to our partners.

The code to connect to partner's host over SMB protocol is  setup in groovy language, utilizing smbj-0.10.0.jar and
uses **DiskShare.list** method to list files on the destination and latter filter them based on regex pattern.

Post we upgraded the JAR, the code didn't work as is. I read around the forums for related issue and picked about setting AccessMask and Dialects #652  - to make it work. Tried different combination but still continue to get errors as captured below:

Please help provide - workaround, config changes - to overcome this issue.

**Error Message being received:**

Using smbj-0.11.1.jar
--withDialects(SMB2Dialect.SMB_3_0)

1) Access Mask set as MAXIMUM_ALLOWED
`Error  : STATUS_PRIVILEGE_NOT_HELD (0xc0000061): Create failed for \\10.*.*.*\***********`

2) Access Mask set as GENERIC_WRITE
`com.hierynomus.mssmb2.SMBApiException: STATUS_ACCESS_DENIED (0xc0000022): Error closing connection to \\10.*.*.*\*$`

3) Access Mask set as FILE_READ_DATA
`com.hierynomus.mssmb2.SMBApiException: STATUS_ACCESS_DENIED (0xc0000022): Query directory failed for SMB2FileId{persistentHandle=53 8d 1b 2c e6 df 45 58}`

![img](https://user-images.githubusercontent.com/60449731/198149593-85af5b80-39ff-4b3c-8ea7-7295c2e276c9.jpg)

Below is the snippet of the code:

```
import java.lang.*;	
import com.hierynomus.smbj.SMBClient;
import com.hierynomus.smbj.SmbConfig;
import com.hierynomus.smbj.auth.AuthenticationContext;
import com.hierynomus.smbj.session.Session;
import com.hierynomus.smbj.share.DiskShare;
import com.hierynomus.msfscc.fileinformation.FileIdBothDirectoryInformation;
import com.hierynomus.smbj.connection.Connection;
import com.hierynomus.mssmb2.SMBApiException;
import java.util.regex.*;

import com.hierynomus.mssmb2.SMB2Dialect;
import com.hierynomus.security.bc.BCSecurityProvider;
import com.hierynomus.msdtyp.AccessMask;
import java.util.EnumSet;
String log="";
Integer ExceptionCheck = 1 ;
exchProperties = context.exchange.properties
try
{
```
	SmbConfig clientConfig = SmbConfig.builder()
        .withDfsEnabled(true)
        .withSecurityProvider(new BCSecurityProvider())
        .withDialects(SMB2Dialect.SMB_3_0)
        .withMultiProtocolNegotiate(true)
        .withSigningRequired(true)
        .build();`
	
	SMBClient client = new SMBClient(clientConfig);
	String Host_Shared = SOURCE_FILE_SFTP_HOST;
	String[] host_shared = Host_Shared.split("\\\\");
	String Host = host_shared[2];
	InetAddress inetAddress = InetAddress.getByName(Host);
    String ipAddress = inetAddress.getHostAddress();
	String DiskDirectory = host_shared[3];
	Connection connection = client.connect(ipAddress); 
   	AuthenticationContext authContext = new AuthenticationContext(SOURCE_FILE_SFTP_USERNAME, SOURCE_FILE_SFTP_PASSWORD.toCharArray(), DOMAIN);
   	Session session = connection.authenticate(authContext);
	if(session == null) {return;}
	DiskShare share = (DiskShare) session.connectShare(DiskDirectory);
	exchProperties = context.exchange.properties
	Pattern regex;
	Integer MatchCount = 0 ;
	Integer EmptyCheck = 1 ;
	String allRegexesFromTR=FILE_NAME_PATTERN; 
	regex = Pattern.compile(allRegexesFromTR,Pattern.CASE_INSENSITIVE);  
  	String location = SOURCE_FILE_SFTP_LOCATION
	String fileDirectory = location.substring(1)
	//println fileDirectory;
        String FileContent = "";
	println "before loop";
       
	List<FileIdBothDirectoryInformation> files = (List<FileIdBothDirectoryInformation>) share.list(fileDirectory, FileIdBothDirectoryInformation.class, null, EnumSet.of(AccessMask.MAXIMUM_ALLOWED));
	//Code keeps failing in above step
	
	for (FileIdBothDirectoryInformation file : files) 
	{
		println "Entering Loop";
                String filename = file.getFileName();
		println filename ;	
                        if(filename .length()<3) 
			{
				continue; 
			}
			.........................................
			.........................................
			.........................................
	}
```
catch( SMBApiException ioe ) 
{
	log+=ioe;
	println ioe.getMessage();
        exchProperties.put("Log",log);
	exchProperties.put("isException",ExceptionCheck);
	context.createExchangeResponse(exchProperties,"")
}
```

**JARs used:**
smbj-0.11.1.jar
asn-one-0.5.0.jar
mbassador-1.3.0.jar
bcprov-jdk15on-1.60.jar
slf4j-api-1.7.32.jar

**Comment by himanshuccl** [comment]
We were able to eventually resolve this by opting for below modifications:

1)  We set the Access mask as - FILE_LIST_DIRECTORY.

`List<FileIdBothDirectoryInformation> files = (List<FileIdBothDirectoryInformation>) share.list(fileDirectory, FileIdBothDirectoryInformation.class, null, EnumSet.of(AccessMask.FILE_LIST_DIRECTORY));`


2) On SMB3 host - we where also getting Authentication denied as reported here : #606 
This we resolved, by setting  in config option this - .withEncryptData(true)

We also chose to set multiple dialect under the config.

`SmbConfig clientConfig = SmbConfig.builder()
    .withDfsEnabled(true)
    .withSecurityProvider(new BCSecurityProvider())
    .withDialects(SMB2Dialect.SMB_2XX, SMB2Dialect.SMB_2_0_2, SMB2Dialect.SMB_2_1, SMB2Dialect.SMB_3_0)
    .withMultiProtocolNegotiate(true)
    .withSigningRequired(true)
    .withEncryptData(true)
    .build();`


Thanks.
